### DISCUSSION TITLE
How to merge and transfer 2k sui coin objects

### DISCUSSION THREAD

**Discussion by igdeman** [MystenLabs/sui#18254] [Q&A] [answered]
Hi there,

I have a situation where I have an address with 2,000 SUI coin objects. I was wondering how I can merge all these coin objects and transfer them to another address. Is it possible to do this through a single transaction, or do I need to make multiple transactions to merge them first and then transfer them at the end? I'm interested in a solution using the TypeScript SDK. Any ideas?

**Comment by igdeman** [comment, marked as answer]
I have come up with a solution, but I'm not sure if it is optimal:

```
it("Testing Transfer All Coins", async () => {
    const coins = await getCoins(sender.getPublicKey().toSuiAddress(), client);
    const gasCoins = coins.splice(0, Math.min(50, coins.length));
    const transferCoins: CoinStruct[] = [];

    const tx = new Transaction();
    tx.setSender(sender.getPublicKey().toSuiAddress());
    while (coins.length > 0) {
      const chunk = coins.splice(0, Math.min(512, coins.length));
      const [mergedCoin] = chunk.splice(0, 1);
      if (chunk.length > 0) tx.mergeCoins(mergedCoin.coinObjectId, chunk.map(coin => coin.coinObjectId));
      transferCoins.push(mergedCoin);
    }
    tx.setGasPayment(gasCoins.map(coin => {
      return {
        objectId: coin.coinObjectId,
        version: coin.version,
        digest: coin.digest,
      }
    }))
    tx.transferObjects([tx.gas, ...transferCoins.map(coin => coin.coinObjectId)], receiver)
    const { bytes, signature } = await sender.signTransaction(await tx.build({ client }));
    expect(await sender.getPublicKey().verifyTransaction(Buffer.from(bytes, "base64"), signature)).toBeTruthy();

    const simulationResult = await client.dryRunTransactionBlock({ transactionBlock: Buffer.from(bytes, "base64") });
    expect(simulationResult.effects.status.status).toBe("success");
  })
```

this line
`const coins = await getCoins(sender.getPublicKey().toSuiAddress(), client);`
recursively retrieves all coins

**Comment by Eis-D-Z** [comment]
There are limits you might hit. For gas coins the limit is 256 so you'd have to do this 10 times, for objects inputs its 1024.

So I would create a generic function that accepts a list  of up to 1000 coins and merges them normally, like take the first coin from the array and merge the rest into them and then just send the coin to the same address. Then I'd call this function twice but with "await" so make sure you get a response from the first one to call the second. Lastly after all 2000 objects are merged into 2 objects, assuming you want to empty this address I'd create a PTB that transfers tx.gas to the new address. When called like that available coins will be merged (since 2 < 256).
